fix for 9699. Added bimap for language-lcid mapping

yenniejunvu opened the pull request:
Fixes #9699

Added a private static bimap for supporting more lcids/languages when exporting to MS Office 2007. 
Added simple unit tests, but not sure if those are necessary.


<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

```[tasklist]
### Compulsory checks
- [x ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.
```

koppor commented on src/main/java/org/jabref/logic/msbib/MSBibMapping.java:
```diff
@@ -24,6 +24,45 @@ public class MSBibMapping {
 
     private static final HashBiMap<Field, String> BIBLATEX_TO_MS_BIB = HashBiMap.create();
 
+    private static final HashBiMap<String, Integer> LANG_TO_LCID = HashBiMap.create();
```
Do you really need `HashBiMap` here? Isn't BiMap enough?

See https://github.com/HugoMatilla/Effective-JAVA-Summary#52-refer-to-objects-by-their-interface for a longer explanation

yenniejunvu replied:
Doesnt the same logic apply for BIBLATEX_TO_MS_BIB as well?

koppor replied:
I think so.

koppor commented on src/main/java/org/jabref/logic/msbib/MSBibMapping.java:
```diff
@@ -131,8 +170,8 @@ public static MSBibEntryType getMSBibEntryType(EntryType bibtexType) {
      */
     public static int getLCID(String language) {
         // TODO: add language to LCID mapping
```
This TODO can be removed.

Also fix the checkstyle issue!

koppor reviewed (changes requested):
Good initial shot.

[MSBibConverter.java](https://github.com/JabRef/jabref/blob/1896c99201b299fbfd88545a2a27a37e462490a6/src/main/java/org/jabref/logic/msbib/MSBibConverter.java#L74-L74) also should be removed.

Please add test for MsBibConverter using <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>T</kbd>

![image](https://user-images.githubusercontent.com/1366654/229543668-5692d44c-d920-4762-baed-9a3cb9592654.png)

Press then <kbd>Enter</kbd>

Then create a test case with a BibEntry.

See [IEEETest.java](https://github.com/JabRef/jabref/blob/1896c99201b299fbfd88545a2a27a37e462490a6/src/test/java/org/jabref/logic/importer/fetcher/IEEETest.java#L28-L28) for a more complete example.

koppor commented:
See also that the tests start to fail: Please check the tests:

![image](https://user-images.githubusercontent.com/1366654/229546655-32e33544-0b1e-4486-9ba5-9d6ed2dee945.png)

yenniejunvu commented:
My bad about the unit tests. I'll take a look and make some fixes. Thanks for the quick feedback.

yenniejunvu commented:
`@return Returns 0 for English`

In regards to the unit tests failing, 
I saw this comment above the `getLCID` function while trying to fix the issue, and noticed that if I set the LCID to '0' in the BiMap for english, the tests pass fine. However, according to documentation,
"Word sets LCID to 0 to mean the value should be  determined at runtime by the application." (https://learn.microsoft.com/en-us/openspecs/office_standards/ms-oe376/6c085406-a698-4e12-9d4d-c3b0ee3dbc4a)

Not quite sure what to do here. 

Additionally, since @Alexandra-Stath was assigned to this problem first, you do not need to review this PR until she submits a PR as well or says it is okay. I just do not want this to get merged, since she was technically assigned it first as a course project!

koppor commented:
Without thinking much, my question would be, why can't `0` be kept?`

Are all these null pointer exceptions really because of this?

Siedlerchr commented:
I think the tests are failing because the xml files for the tests probably contain 0 for the language. And there is no map entry with 0 inside. So either add an entry for 0 or do something like getOrDefault

koppor commented on src/test/java/org/jabref/logic/msbib/MsBibMappingTest.java:
```diff
@@ -0,0 +1,19 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class MsBibMappingTest {
```
Please add a test for a number not appearing in the map when executing `MSBibMapping.getLanguage`. You will never know, which language Microsoft will add in the future. The converter should also be able to handle unknown language numbers (with a best-effort?)

yenniejunvu commented:
@Siedlerchr I'm not sure if that's the case, because the previous implementation of getLCID() returned the hexidecimal value of 1033, which is the LCID for american english.

koppor commented:
> @Siedlerchr I'm not sure if that's the case, because the previous implementation of getLCID() returned the hexidecimal value of 1033, which is the LCID for american english.

This is good!

Without thinking deeper about the consequences, just fallback to `1033` if there is no mapping in the map?

(And please adapt the wrong comment `@return Returns 0 for English` if not already done so)

Siedlerchr commented:
Don't guess, check the files! https://github.com/JabRef/jabref/tree/6f9eaafce606e61426565f5316e8a0b6f68bf227/src/test/resources/org/jabref/logic/exporter
and https://github.com/JabRef/jabref/tree/6f9eaafce606e61426565f5316e8a0b6f68bf227/src/test/resources/org/jabref/logic/importer/fileformat

yenniejunvu commented:
I added MSBibConverterTest but I am not sure if that is what you guys are expecting.
Additionally, I added some error handling. If there is an invalid language/lcid passed, it all defaults to english.

Siedlerchr commented on src/main/java/org/jabref/logic/msbib/MSBibMapping.java:
```diff
@@ -163,24 +163,28 @@ public static MSBibEntryType getMSBibEntryType(EntryType bibtexType) {
+        if (!LANG_TO_LCID.containsKey(language)) {
+            return 1033;
+        }
         return LANG_TO_LCID.get(language);
```
You can use getOrDefault

yenniejunvu commented:
Not sure why that one test is now suddenly failing, it works fine on my local machine...

Siedlerchr commented on src/main/java/org/jabref/logic/msbib/MSBibMapping.java:
```diff
@@ -22,7 +23,46 @@ public class MSBibMapping {
+
+    private static final BiMap<String, Integer> LANG_TO_LCID = HashBiMap.create();
+
+    static {
```
Can you please add a comment with the link to the mapping so someone else knows where this comes from.
Otherwise it looks good to me now
https://learn.microsoft.com/en-us/openspecs/office_standards/ms-oe376/6c085406-a698-4e12-9d4d-c3b0ee3dbc4a

Siedlerchr commented:
The test sometimes fails randomly on CI. Don't worry. If you think your PR is ready then you can mark it as ready for review. In my opinion it looks good now

koppor commented on CHANGELOG.md:
```diff
@@ -381,7 +381,7 @@ Note that this project **does not** adhere to [Semantic Versioning](http://semve
 - We improved the Drag and Drop behavior in the "Customize Entry Types" Dialog [#6338](https://github.com/JabRef/jabref/issues/6338)
 - When determining the URL of an ArXiV eprint, the URL now points to the version [#8149](https://github.com/JabRef/jabref/pull/8149)
-- We Included all standard fields with citation key when exporting to Old OpenOffice/LibreOffice Calc Format [#8176](https://github.com/JabRef/jabref/pull/8176)
+- We Included all standard fields with citation key when exporting to Old OpenOffice/LibreExpoa Calc Format [#8176](https://github.com/JabRef/jabref/pull/8176)
```
@Siedlerchr What is "LibreExpoa"? Is this change in CHANGELOG.md really intended?

Siedlerchr replied:
I was editing this on the github, this was not intended

koppor replied:
Fixed it directly in `main` at https://github.com/JabRef/jabref/commit/461e915105dfb5acd53b917b541a8a93026b8dad
